tsl/src/remote/copy_fetcher.c

```diff
@@ -338,85 +385,101 @@ copy_fetcher_complete(CopyFetcher *fetcher)
 			const int16 natts = copy_data_read_int16(&copy_data);
 			if (natts == -1)
 			{
-				fetcher->state.eof = true;
-				break;
-			}
+				/*
+				 * From the PostgreSQL (libpq) docs: The file trailer consists
+				 * of a 16-bit integer word containing -1. This is easily
+				 * distinguished from a tuple's field-count word. A reader
+				 * should report an error if a field-count word is neither -1
+				 * nor the expected number of columns. This provides an extra
+				 * check against somehow getting out of sync with the data.
+				 */
+				fetcher->file_trailer_received = true;
```

Comment:
Why do we need separate logic for the file trailer as compared to other kinds of eof? We do the same thing for them anyway, right?

Reply:
The file trailer and EOF are technically different things.

The file trailer _is_ data in the stream. EOF is the return value from `PGgetCopyData()`. I imagine that one could in theory send two or more files in the same COPY stream, delineated by a file trailer. EOF indicates end of the stream. These are technically two different things.

I prefer to treat the file trailer as data, and not do special casing for what to do in that case. Instead, just continue reading in the loop until `PGgetCopyData()` return -1 as specified by the protocol. Whether we read EOF  here or in the next iteration of the loop doesn't matter technically, but it seems more elegant to me to have `PGgetCopyData()` in  only one place.

Reply:
> I imagine that one could in theory send two or more files in the same COPY stream, delineated by a file trailer.

This doesn't happen and it's not something that we should try to support. I agree that we should add a check that the file trailer is followed by EOF. This means that in practice, 'eof' and 'trailer' are the same thing.

> it seems more elegant to me to have PQgetCopyData() in only one place.

What's wrong with adding one call and checking that it returns EOF? Instead of that, we have two flags with barely distinguishable meaning, both of which we have to check in several other places. Basically we're using the difference between these flags to make additional iteration of the loop, which should only call `PQgetCopyData`. To me this looks confusing and unnecessarily complicated.

Reply:
> What's wrong with adding one call and checking that it returns EOF? Instead of that, we have two flags with barely distinguishable meaning, both of which we have to check in several other places. Basically we're using the difference between these flags to make additional iteration of the loop, which should only call `PQgetCopyData`. To me this looks confusing and unnecessarily complicated.

I guess there are pros and cons with both approaches. But the cons are bigger in my book.

The flags effectively express the  state machine that the fetcher implements. Perhaps it makes sense to turn them into an enum: RECEIVING, FILE_TRAILER, EOF, or similar, instead of separate booleans in case you are concerned about multiple flags. I see nothing wrong with explicitly expressing the state machine. I imagine that for many that would actually bring clarity in what's going on as opposed to it being implicit.

The main reason I don't think your suggest approach is great, however,  is because it requires checking for not only the expected condition in both places where `PQgetCopyData()` is called, but also handle the conditions that are not expected. `PQgetCopyData()` returns three states represented by the values  >=0, -1, -2 for receiving data, EOF, and error, respectively. If I check for those conditions in one place, all of them will likely be executed at some point. But, if I split the handling of those states across different places in the code, I need to check for the "expected" condition but also add error handling for the unexpected ones. The unexpected condition is likely to never going to be executed, with more dead code as a result and worse CodeCov.

Further, if I read EOF as a special case after receiving the file trailer, I also need to break out of the loop. But that means I need to duplicate some code to free memory read for the file trailer which is normally done at the end of the loop (thus having `PQmemfree()` in two places as well). I also need to modify `fetcher->state.num_tuples` in two places to account for having received a legitimate data tuple vs a file trailer.

To summarize, I guess there are pros and cons with both approaches, although I personally don't like special-casing like you suggest. It is, IMO, cleaner with a general loop where you handle _all_ return values of a function in one place. Right now I am not convinced by your arguments for this change that has no functional improvement. It will only increase the amount of similar code and also dead code.

I suggest keeping this as is unless the next reviewer also feels strongly that we should change it.

Reply:
Ok, I'll consider this change for a refactoring PR as a follow up.